types/msgservice/validate.go

```diff
@@ -0,0 +1,116 @@
+
+	msg "cosmossdk.io/api/cosmos/msg/v1"
+)
+
+// ValidateAnnotations validates that the proto annotations are correct.
+// More specifically, it verifies:
+// - all services named "Msg" have `(cosmos.msg.v1.service) = true`,
+// - all sdk.Msgs have correct `(cosmos.msg.v1.signer) = "..."`.
+//
+// More validations can be added here in the future.
+//
+// If `protoFiles` is nil, then protoregistry.GlobalFile will be used.
+func ValidateProtoAnnotations(protoFiles *protoregistry.Files) error {
+	if protoFiles == nil {
+		protoFiles = protoregistry.GlobalFiles
+	}
+
+	var (
+		serviceErrs []error
+		messageErrs []error
+	)
+	protoFiles.RangeFiles(func(fd protoreflect.FileDescriptor) bool {
+		for i := 0; i < fd.Services().Len(); i++ {
+			sd := fd.Services().Get(i)
+			if sd.Name() == "Msg" {
+				// We use the heuristic that services name Msg are exactly the
+				// ones that need the proto annotations check.
+				err := validateMsgServiceAnnotations(protoFiles, sd)
+				if err != nil {
+					serviceErrs = append(serviceErrs, err)
+				}
+
+				for j := 0; j < sd.Methods().Len(); j++ {
+					err := validateMsgAnnotations(protoFiles, sd.Methods().Get(j).Input())
+					if err != nil {
+						messageErrs = append(messageErrs, err)
+					}
+				}
+			}
+		}
+
+		return true
+	})
+
+	return errors.Join(errors.Join(serviceErrs...), errors.Join(messageErrs...))
+}
+
+// validateMsgServiceAnnotations validates that the service has the
+// `(cosmos.msg.v1.service) = true` proto annotation.
+func validateMsgServiceAnnotations(protoFiles *protoregistry.Files, sd protoreflect.ServiceDescriptor) error {
+	ext := proto.GetExtension(sd.Options(), msg.E_Service)
+	isService, ok := ext.(bool)
+	if !ok {
+		return fmt.Errorf("expected bool, got %T", ext)
+	}
+
+	if !isService {
+		return fmt.Errorf("service %s does not have cosmos.msg.v1.service proto annotation", sd.FullName())
+	}
+
+	return nil
+}
+
+// validateMsgAnnotations validates the sdk.Msg service has the correct
+// `(cosmos.msg.v1.signer) = "..."` proto annotation.
+func validateMsgAnnotations(protoFiles *protoregistry.Files, md protoreflect.MessageDescriptor) error {
+	ext := proto.GetExtension(md.Options(), msg.E_Signer)
+	signers, ok := ext.([]string)
+	if !ok {
+		return fmt.Errorf("expected []string, got %T", ext)
+	}
+
+	if len(signers) == 0 {
+		return fmt.Errorf("sdk.Msg %s does not have cosmos.msg.v1.signer proto annotation", md.FullName())
+	}
+
+	for i, signer := range signers {
+		if signer == "" {
+			return fmt.Errorf("sdk.Msg %s signer at index %d is empty", md.FullName(), i)
+		}
+
+		// Make sure the signer annotation is a correct field of type string
+		fd := md.Fields().ByName(protoreflect.Name(signer))
+		if fd == nil {
+			return fmt.Errorf("sdk.Msg %s has incorrect signer %s", md.FullName(), signer)
+		}
+
+		// The signer annotation should point to:
+		// - either be a string field,
+		// - or a message field who recursively has a "signer" string field.
```

Comment:
would this be a nested signer field? do we use such things in the sdk currently?

Reply:
Yeah - `MsgMultiSend`. But I think that's the only use case. It's somewhat pathological IMHO but we explored killing it and people objected...

Reply:
dam, bank v2 will be our chance 😉